src/main/scala/zio/ftp/SFtp.scala

```diff
@@ -0,0 +1,202 @@
+          .orDie
+    )
+  }
+
+  private[this] def setIdentity(identity: SftpIdentity, username: String)(ssh: SSHClient): Unit = {
+    def bats(array: Array[Byte]): String = new String(array, "UTF-8")
+
+    def initKey(f: OpenSSHKeyFile => Unit): Unit = {
+      val key = new OpenSSHKeyFile
+      f(key)
+      ssh.authPublickey(username, key)
+    }
+
+    val passphrase =
+      identity.privateKeyFilePassphrase.map(pass => PasswordUtils.createOneOff(bats(pass).toCharArray)).orNull
+
+    identity match {
+      case id: RawKeySftpIdentity =>
+        initKey(_.init(bats(id.privateKey), id.publicKey.map(bats).orNull, passphrase))
+      case id: KeyFileSftpIdentity =>
+        initKey(_.init(new File(id.privateKey), passphrase))
+    }
+  }
+
+  def listFiles(
+    basePath: String,
+    predicate: FtpFile => Boolean = _ => true
+  ): ZStream[SFtpClient, IOException, FtpFile] = StreamAccess.accessM[SFtpClient] { client =>
+    val path = if (!basePath.isEmpty && basePath.head != '/') s"/$basePath" else basePath
+    val filter = new RemoteResourceFilter {
+      override def accept(r: RemoteResourceInfo): Boolean = predicate(SftpFileOps(r))
+    }
+
+    ZStream
+      .fromEffect(taskIO(client.ls(path, filter)).catchSome {
+        case ex: SFTPException if ex.getStatusCode == Response.StatusCode.NO_SUCH_FILE =>
+          ZIO.succeed(scala.collection.mutable.Buffer.empty[RemoteResourceInfo])
+      })
+      .flatMap(Stream.fromIterable)
+      .flatMap { f =>
+        if (f.isDirectory) listFiles(f.getPath, predicate)
+        else Stream(SftpFileOps(f))
+      }
+  }
+
+  def stat(path: String): ZIO[SFtpClient, IOException, Option[FtpFile]] =
+    ZIO.accessM[SFtpClient] { client =>
+      taskIO(client.stat(path)).either
+        .map(r => r.map(r => SftpFileOps(path, r)).toOption)
+    }
+
+  def readFile[E](path: String, chunkSize: Int = 2048): ZStream[SFtpClient, IOException, Byte] =
+    StreamAccess.accessM[SFtpClient] { client =>
+      for {
+        remoteFile <- ZStream.fromEffect(
+                       taskIO(client.open(path))
+                         .orDieWith(ex => new IOException(s"File does not exist $path", ex))
+                     )
+
+        is: java.io.InputStream = new remoteFile.ReadAheadRemoteFileInputStream(64) {
+          override def close(): Unit =
+            try {
+              super.close()
+            } finally {
+              remoteFile.close()
+            }
+        }
+
+        input <- Stream.fromInputStream(is, chunkSize).chunks.flatMap(Stream.fromChunk)
+      } yield input
+    }
+
+  def rm(path: String): ZIO[SFtpClient, IOException, Unit] = ZIO.accessM[SFtpClient] { client =>
+    taskIO(client.rm(path))
+      .orDieWith(ex => new IOException(s"Path is invalid. Cannot delete file : $path", ex))
+  }
+
+  def rmdir(path: String): ZIO[SFtpClient, IOException, Unit] = ZIO.accessM[SFtpClient] { client =>
+    taskIO(client.rmdir(path))
```

Comment:
Note that when we are dealing with blocking APIs, we should be sure to lock the tasks to the blocking thread pool.